### ISSUE TITLE
Phase 48.4: TaskDetector — Task Boundary Detection, Distribution Shift & Concept Drift Monitoring

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#941]
### Overview & Motivation

Most continual learning methods assume task boundaries are explicitly provided — the system is told when one task ends and another begins. In real-world deployment, **task boundaries are often unknown or gradual**. The TaskDetector addresses this by automatically detecting distribution shifts, concept drift, and task transitions, enabling truly autonomous continual learning without manual task labels.

This component bridges the gap between controlled experimental settings and real-world deployment by providing reliable task boundary detection and distribution shift monitoring.

### Architecture Components

```
TaskDetector
├── DistributionShiftMonitor
│   ├── MMDDetector                  # Maximum Mean Discrepancy test
│   ├── KLDivergenceTracker          # KL divergence on feature distributions
│   ├── WassersteinDistanceMonitor   # Earth mover's distance tracking
│   ├── KolmogorovSmirnovTest        # KS test for distributional change
│   └── WindowedStatisticsTracker    # Sliding window summary statistics
├── ConceptDriftDetector
│   ├── ADWIN                        # Adaptive Windowing (Bifet & Gavaldà, 2007)
│   ├── PageHinkleyDetector          # Page-Hinkley change detection test
│   ├── DDMDetector                  # Drift Detection Method (Gama et al., 2004)
│   ├── EDDM                         # Early DDM with warning/drift levels
│   └── CUSUM                        # Cumulative sum control chart
├── RepresentationDriftAnalyzer
│   ├── ActivationDistributionMonitor # Track hidden layer activation distributions
│   ├── GradientDirectionTracker     # Detect gradient direction changes
│   ├── LossLandscapeMonitor         # Track loss surface geometry changes
│   ├── FeatureCorrelationTracker    # Monitor inter-feature correlations
│   └── RepresentationSimilarity     # CKA/CCA between old and new representations
├── TaskBoundaryEstimator
│   ├── ChangePointDetector          # Bayesian online changepoint detection
│   ├── OnlineClusteringDetector     # Detect new clusters in feature space
│   ├── ThresholdCalibrator          # Adaptive threshold for boundary decisions
│   ├── MultiSignalFusion            # Combine multiple drift signals
│   └── BoundaryConfidenceEstimator  # Confidence score for detected boundaries
├── CurriculumAnalyzer
│   ├── TaskSimilarityEstimator      # Measure similarity between detected tasks
│   ├── DifficultyEstimator          # Estimate relative task difficulty
│   ├── OrderingOptimizer            # Suggest optimal task ordering
│   └── TransferPotentialPredictor   # Predict beneficial transfer between tasks
└── AlertingSystem
    ├── DriftAlertManager            # Alert when significant drift detected
    ├── GradualVsSuddenClassifier    # Classify drift as gradual or sudden
    ├── SeverityEstimator            # Estimate magnitude of distribution shift
    └── ActionRecommender            # Recommend CL strategy based on drift type
```

### Key Algorithms & Data Structures

**ADWIN (Adaptive Windowing):**
```python
class ADWIN:
    def __init__(self, delta=0.002):
        self.delta = delta
        self.window = deque()

    def update(self, value):
        self.window.append(value)
        while len(self.window) > 1:
            # Test all possible splits
            for split in self._possible_cuts():
                w0, w1 = self._split_window(split)
                epsilon = self._hoeffding_bound(len(w0), len(w1))
                if abs(mean(w0) - mean(w1)) >= epsilon:
                    # Drift detected — drop old portion
                    self._shrink_window(split)
                    return DriftSignal(detected=True, magnitude=abs(mean(w0)-mean(w1)))
        return DriftSignal(detected=False)
```

**Bayesian Online Changepoint Detection (Adams & MacKay, 2007):**
```python
class BayesianChangePointDetector:
    def update(self, x):
        # Compute predictive probability under each run length
        pred_probs = self.hazard_function.predict(x, self.run_lengths)
        # Growth probability: P(r_t = r_{t-1} + 1)
        growth = self.joint_prob * pred_probs * (1 - self.hazard_rate)
        # Changepoint probability: P(r_t = 0)
        changepoint = (self.joint_prob * pred_probs * self.hazard_rate).sum()
        # Update run length distribution
        self.joint_prob = torch.cat([changepoint.unsqueeze(0), growth])
        self.joint_prob /= self.joint_prob.sum()
        # Detect if changepoint probability exceeds threshold
        return self.joint_prob[0] > self.threshold
```

**MMD Two-Sample Test:**
```python
class MMDDetector:
    def compute_mmd(self, X, Y, kernel="rbf", bandwidth=1.0):
        K_XX = self.kernel(X, X, bandwidth)
        K_YY = self.kernel(Y, Y, bandwidth)
        K_XY = self.kernel(X, Y, bandwidth)
        mmd = K_XX.mean() + K_YY.mean() - 2 * K_XY.mean()
        return mmd

    def test_drift(self, reference_window, current_window, alpha=0.05):
        mmd = self.compute_mmd(reference_window, current_window)
        threshold = self.permutation_test_threshold(reference_window, current_window, alpha)
        return DriftResult(mmd=mmd, threshold=threshold, drift_detected=mmd > threshold)
```

### API Surface

```python
class TaskDetector:
    def __init__(self, config: TaskDetectorConfig): ...
    def update(self, features: Tensor, loss: float = None) -> DetectionResult: ...
    def detect_drift(self, method: str = "ensemble") -> DriftReport: ...
    def get_task_boundaries(self) -> List[TaskBoundary]: ...
    def estimate_task_similarity(self, task_a: str, task_b: str) -> float: ...
    def get_drift_severity(self) -> DriftSeverity: ...
    def classify_drift_type(self) -> DriftType: ...  # gradual, sudden, recurring, incremental
    def recommend_strategy(self, drift_report: DriftReport) -> StrategyRecommendation: ...
    def set_reference_distribution(self, features: Tensor) -> None: ...
    def get_monitoring_dashboard(self) -> MonitoringState: ...
```

### Testing Strategy

- Synthetic drift injection: verify detection on known changepoints (step function, gradual ramp)
- ADWIN: validate correct window shrinking on synthetic streams with known drift points
- MMD: verify type-I error rate at specified alpha on same-distribution samples
- Bayesian changepoint: verify run length posterior concentrates at true changepoint
- Multi-signal fusion: verify combined detection outperforms individual detectors
- Latency test: time-to-detection on sudden vs gradual drift
- False positive rate monitoring on stationary distributions

### References

1. van de Ven & Tolias (2019) — *Three scenarios for continual learning*
2. Adams & MacKay (2007) — *Bayesian Online Changepoint Detection*
3. Bifet & Gavaldà (2007) — *Learning from Time-Changing Data with Adaptive Windowing* (ADWIN)
4. Gama et al. (2004) — *Learning with Drift Detection* (DDM)
5. Gretton et al. (2012) — *A Kernel Two-Sample Test* (MMD)
6. Lu et al. (2019) — *Learning under Concept Drift: A Review*
7. De Lange et al. (2021) — *A continual learning survey*

Part of **Phase 48: Continual Learning & Catastrophic Forgetting Prevention**
Discussion: #937
